lib/TPP/TileConsumerAndFuseProducers.cpp

```diff
@@ -14,119 +14,421 @@
   SmallVector<Range> iterationDomain =
-      cast<TilingInterface>(linalgOp.getOperation())
-          .getIterationDomain(builder);
-  for (const auto &it : llvm::enumerate(iterationDomain)) {
-    // fine, we are not tiling along this dimension.
-    if (tiles[it.index()] == 0)
+      cast<TilingInterface>(op).getIterationDomain(builder);
+  SmallVector<utils::IteratorType> loopIteratorTypes =
+      cast<TilingInterface>(op).getLoopIteratorTypes();
+  if (tileSizes.size() > iterationDomain.size())
+    return false;
+  for (size_t tileIdx = 0, idxEnd = tileSizes.size(); tileIdx < idxEnd;
+       tileIdx++) {
+    // Allow tiling only on parallel loop iterators.
+    if (!linalg::isParallelIterator(loopIteratorTypes[tileIdx]))
+      return false;
+    std::optional<int64_t> sizeRangeAtIdx =
+        getSizeRangeAtIdx(iterationDomain[tileIdx]);
+    // Non constant range. Bail out and do not add the op as candidate.
+    if (!sizeRangeAtIdx)
+      return false;
+    // Skip tile equals 0. A '0' tile means skip this dimension.
+    if (tileSizes[tileIdx] == 0)
       continue;
-    // require static loop range
-    if (!isStaticRange(it.value()))
-      return failure();
-    int64_t sizeRange = getSizeRange(it.value());
-    // fail if the tail size equals the range
-    // or is not a full tile.
-    if (tiles[it.index()] == sizeRange)
-      return failure();
-    if (sizeRange % tiles[it.index()] != 0)
-      return failure();
+    // Fail if the tile size equals the range or it is not a full tile.
+    if (tileSizes[tileIdx] == *sizeRangeAtIdx ||
+        *sizeRangeAtIdx % tileSizes[tileIdx] != 0)
+      return false;
+  }
+  // Candidate op is good to go.
+  return true;
+}
+
+// Returns a bit vector of size number of loops of the `interfaceOp` with
+// the bits corresponding to outer parallel loops set to `true`.
+// Adapted from IREE: `FormDispatchRegions.cpp`.
+static llvm::SmallBitVector getOuterParallelLoops(Operation *op) {
+  auto interfaceOp = dyn_cast<TilingInterface>(op);
+  if (!interfaceOp)
+    return llvm::SmallBitVector{};
+
+  SmallVector<utils::IteratorType> loopIteratorTypes =
+      interfaceOp.getLoopIteratorTypes();
+  llvm::SmallBitVector parallelLoops(loopIteratorTypes.size());
+  for (auto iteratorType : llvm::enumerate(loopIteratorTypes)) {
+    if (iteratorType.value() != utils::IteratorType::parallel)
+      break;
+    parallelLoops.set(iteratorType.index());
   }
-  return success();
+  return parallelLoops;
+}
+
+// Return the tile sizes as bit vector.
+static llvm::SmallBitVector
+getTileBitVectorConfig(ArrayRef<int64_t> tileSizes, size_t rootConsumerParallelLoops) {
+  assert(tileSizes.size() <= rootConsumerParallelLoops);
+  llvm::SmallBitVector tileConfig(rootConsumerParallelLoops, false);
+  for (size_t idx : llvm::seq<size_t>(0, tileSizes.size()))
+    if (tileSizes[idx] != 0)
+      tileConfig.set(idx);
+  return tileConfig;
+}
+
+static void _and(llvm::SmallBitVector &parallelLoops,
```

Comment:
I think you could just use the built-in `SmallBitVector &operator&=` instead. Or at least give this function a bit more expressive name.